app/src/main/java/com/example/composetemplate/repositories/FirebaseDataSource.kt

```diff
@@ -1,93 +1,123 @@
 package com.example.composetemplate.repositories
 
+import android.util.Base64
 import com.example.composetemplate.data.models.local_models.User
-import com.example.composetemplate.data.remote.requests.FirebaseUserRequests
+import com.example.composetemplate.managers.TokenData
+import com.example.composetemplate.managers.TokenFetcher
 import com.example.composetemplate.utils.LoginCallback
 import com.example.composetemplate.utils.SuccessCallback
-import com.google.firebase.Firebase
 import com.google.firebase.auth.FirebaseAuth
-import com.google.firebase.auth.auth
+import kotlinx.coroutines.suspendCancellableCoroutine
+import org.json.JSONObject
+import kotlin.coroutines.resume
 
 /**
  * In the MVVM architecture, the DataSource layer is responsible for providing data to the repository
  * from various sources, such as remote APIs, local databases, or in-memory caches.
  * allowing the repository to seamlessly aggregate and supply data to the ViewModel without worrying about the underlying data origins or access mechanisms.
  */
-class FirebaseDataSource() : AuthDBServiceable {
+class FirebaseDataSource : AuthDataSource, TokenFetcher {
```

Comment:
I think we have an architectural issue here. As you explained in the comment, the DataSource is supposed to be responsible for providing data to the repository. However, this class does more than just provide data. For example, actions like signInWithEmailAndPassword or logOut don't seem to align with the role of a DataSource. If you disagree, we can consult with Tom about it.